### PR TITLE
WIP: Add support for rds DBCluster using ACK code gen

### PR DESCRIPTION
**Pull request by hiteshghia-splunk**
### Description of your changes
Trying to add support for resources in RDS  using ACK code gen, starting with db cluster and ignoring every other type for now.
Steps followed till now -  
- Added generator-config.yaml in provider-aws/apis/rds/v1alpha1.
- Ran `./bin/ack-generate crossplane rds --provider-dir /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds` from the ACK repo. (PR https://github.com/aws/aws-controllers-k8s/pull/583and cherry picked https://github.com/aws/aws-controllers-k8s/pull/586 on top)
- Added empty struct for `CustomDBClusterParameters` in  generated zz_db_cluster.go and ran `make generate`.

Fixes https://github.com/crossplane/provider-aws/issues/272

I have:

- [x] Read and followed Crossplane's [contribution process].
- [] Run `make reviewable test` to ensure this PR is ready for review.

### CODE REVIEW

**Comment by hiteshghia-splunk** [comment]
After making the changes in the description above, build fails with the following, looking into it.. - 
`WARN [runner] Can't run linter goanalysis_metalinter: bodyclose: failed prerequisites: [buildssa@github.com/crossplane/provider-aws/pkg/controller/rds/dbcluster: analysis skipped: errors in package: [/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/hooks.go:25:18: could not import github.com/crossplane/crossplane-runtime/apis/core/v1alpha1 (pkg/controller/rds/dbcluster/hooks.go:25:2: module github.com/crossplane/crossplane-runtime@latest found (v0.12.0), but does not contain package github.com/crossplane/crossplane-runtime/apis/core/v1alpha1) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/hooks.go:54:5: cr.SetConditions undefined (type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster has no field or method SetConditions) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:54:12: mg (variable of type resource.Managed) cannot have dynamic type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster (missing method GetCondition) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:71:12: mg (variable of type resource.Managed) cannot have dynamic type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster (missing method GetCondition) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:90:14: resp.Items undefined (type *rds.DescribeDBClustersOutput has no field or method Items) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:104:12: mg (variable of type resource.Managed) cannot have dynamic type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster (missing method GetCondition) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:335:12: mg (variable of type resource.Managed) cannot have dynamic type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster (missing method GetCondition) /Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws-rds/pkg/controller/rds/dbcluster/zz_controller.go:345:12: mg (variable of type resource.Managed) cannot have dynamic type *"github.com/crossplane/provider-aws/apis/rds/v1alpha1".DBCluster (missing method GetCondition)]]
WARN [runner] Can't run linter unused: buildir: failed to load package : could not load export data: no export data for "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
ERRO Running error: buildir: failed to load package : could not load export data: no export data for "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
20:10:03 [FAIL]`

**Comment by muvaf** [comment]
Hi @hiteshghia-splunk ! Thanks for working on this. I believe the current `main` branch of ACK contains what's needed to be cherry-picked for this resource. However, there is one problem there regarding `ReadMany` template which `DBCluster` uses and I'm planning to fix it tomorrow. But until then you can manually fix that in generated code to test your PR.

Here is the most up-to-date instructions that you can use to do the implementation https://github.com/crossplane/provider-aws/pull/486

**Comment by hiteshghia-splunk** [comment]
@muvaf updated this mr with the latest instructions from the code-generation readme.
Need some help on how to deal with the db  cluster statuses in postObserve and if you have  some local testing flow to deploy crossplane master + my version of provider-aws to a kind cluster to iterate and tests changes, that will be awesome.

File: pkg/controller/rds/dbcluster/hooks.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/rds/v1alpha1"
```
Comment (hiteshghia-splunk) [inline pkg/controller/rds/dbcluster/hooks.go:19]: organize imports.

File: CODE_GENERATION.md [hunk 2]
```diff
@@ -47,8 +47,8 @@ ignore:
 ```
 
 When you re-run the generation with this configuration, existing files won't be
-deleted. So you may want to delete everything in `apis/<serviceid>/v1alpha1` except
-`generator-config.yaml` and then re-run the command.
+deleted. So you may want to delete everything in `apis/<serviceid>/v1alpha1` and `pkg/controller/<serviceid>` except
+`generator-config.yaml` and  .... and then re-run the command.
```
Comment (muvaf) [inline CODE_GENERATION.md:51]:
```suggestion
`generator-config.yaml` and then re-run the command.
```

File: CODE_GENERATION.md [hunk 3]
```diff
@@ -157,11 +157,14 @@ in generation and then inject the value of our annotation in its place in the SD
 
 Likely, we need to do this injection before every SDK call. The following is an
 example for hook functions injected:
+// TODO: we mention to create setup.go in the previous step but all existing examples
```
Comment (muvaf) [inline CODE_GENERATION.md:160]: `hooks.go` was generated and then edited by the user but now we don't generate it at all. In the future, we'll likely decrease the amount of code goes in that file but `SetupCRDNAME` function will stay, so I wanted to rename it as `setup.go` to be future-proof.

File: apis/rds/v1alpha1/custom_types.go [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package v1alpha1
```
Comment (muvaf) [inline apis/rds/v1alpha1/custom_types.go:1]: License header is missing.

File: apis/rds/v1alpha1/generator-config.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+ignore:
+  field_paths:
+    - DescribeDBClustersOutput.DBClusterIdentifier
+    - CreateDBClusterInput..DBClusterIdentifier
+    - ModifyDBClusterInput.DBClusterIdentifier
+    - ModifyDBClusterInput.DBClusterIdentifier
```
Comment (muvaf) [inline apis/rds/v1alpha1/generator-config.yaml:6]:
```suggestion
```

Duplicate entry.

File: apis/rds/v1alpha1/zz_db_cluster.go [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+	//
+	//    * Must contain from 1 to 63 letters, numbers, or hyphens.
+	//
+	//    * First character must be a letter.
+	//
+	//    * Can't end with a hyphen or contain two consecutive hyphens.
+	//
+	// Example: my-cluster1
+	// +kubebuilder:validation:Required
+	DBClusterIdentifier *string `json:"dbClusterIdentifier"`
```
Comment (muvaf) [inline apis/rds/v1alpha1/zz_db_cluster.go:75]: This should be omitted by ACK. To do that, you need to add this field in every AWS call in `ignore.field_paths`. I think `DeleteDBCluster` is the missing call that you need to add.

File: apis/rds/v1alpha1/zz_doc.go [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// +kubebuilder:object:generate=true
```
Comment (muvaf) [inline apis/rds/v1alpha1/zz_doc.go:1]: In the latest `main` branch, this file has the necessary license header. I'd suggest pulling latest ACK and run generation again.

File: apis/rds/v1alpha1/zz_db_cluster.go [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+	// The Active Directory directory ID to create the DB cluster in.
+	//
+	// For Amazon Aurora DB clusters, Amazon RDS can use Kerberos Authentication
+	// to authenticate users that connect to the DB cluster. For more information,
+	// see Kerberos Authentication (https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/kerberos-authentication.html)
+	// in the Amazon Aurora User Guide.
+	Domain *string `json:"domain,omitempty"`
+	// Specify the name of the IAM role to be used when making API calls to the
+	// Directory Service.
+	DomainIAMRoleName *string `json:"domainIAMRoleName,omitempty"`
```
Comment (muvaf) [inline apis/rds/v1alpha1/zz_db_cluster.go:111]: Could you implement the referencerse wherever possible? It makes a huge difference in UX for the users.

File: pkg/controller/rds/dbcluster/hooks.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	return obs, nil
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.CreateDBClusterInput) error {
+	obj.DBClusterIdentifier = aws.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func preUpdate(_ context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.ModifyDBClusterInput) error {
```
Comment (muvaf) [inline pkg/controller/rds/dbcluster/hooks.go:72]: Keep in mind that unless `isUpToDate` is overridden, reconciler won't call `Update` at all. From my experience working on `RDSInstance`, I expect the update logic to be tricky, like not allowing every field to be updated with one call etc. But you don't have to deal with update for now, it could be in a separate PR.

**Review by muvaf — COMMENTED** [review]
This is looking pretty good. I left a few comments. @hiteshghia-splunk  have you had the chance to manually test this with a `dbcluster.yaml`?

File: apis/rds/v1alpha1/zz_db_cluster.go [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+	// sent over the wire and is only used for presigning. This value should always
+	// have the same region as the source ARN.
+	SourceRegion *string `json:"sourceRegion,omitempty"`
+	// A value that indicates whether the DB cluster is encrypted.
+	StorageEncrypted *bool `json:"storageEncrypted,omitempty"`
+	// Tags to assign to the DB cluster.
+	Tags []*Tag `json:"tags,omitempty"`
+	// A list of EC2 VPC security groups to associate with this DB cluster.
+	VPCSecurityGroupIDs       []*string `json:"vpcSecurityGroupIDs,omitempty"`
+	CustomDBClusterParameters `json:",inline"`
```
Comment (dan1) [inline apis/rds/v1alpha1/zz_db_cluster.go:331]: hey @hiteshghia-splunk do we need the CustomDBClusterParameters struct defined here?

**Review by dan1 — COMMENTED** [review]
I believe we are missing a referencers.go too.

File: apis/rds/v1alpha1/zz_db_cluster.go (hunk 10 above)
Comment (dan1) [reply apis/rds/v1alpha1/zz_db_cluster.go:331]: nv, I see custom_types.go

**Comment by muvaf** [comment]
@hiteshghia-splunk how do you feel about getting this merged before the feature freeze?

**Comment by hiteshghia** [comment]
closing this in favor of -https://github.com/crossplane/provider-aws/pull/546
